pkg/controllers/machinesync/machine_sync_controller.go

```diff
@@ -99,7 +106,83 @@ func (r *MachineSyncReconciler) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 // Reconcile reconciles CAPI and MAPI machines for their respective namespaces.
+//
+//nolint:funlen
 func (r *MachineSyncReconciler) Reconcile(ctx context.Context, req reconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx, "namespace", req.Namespace, "name", req.Name)
+
+	logger.V(1).Info("Reconciling machine")
+	defer logger.V(1).Info("Finished reconciling machine")
+
+	var mapiMachineNotFound, capiMachineNotFound bool
+
+	// Get the MAPI Machine.
+	mapiMachine := &machinev1beta1.Machine{}
+	mapiNamespacedName := client.ObjectKey{
+		Namespace: r.MAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, mapiNamespacedName, mapiMachine); apierrors.IsNotFound(err) {
+		logger.Info("MAPI Machine not found")
+
+		mapiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get MAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("failed to get MAPI machine: %w", err)
+	}
+
+	// Get the corresponding CAPI Machine.
+	capiMachine := &capiv1beta1.Machine{}
+	capiNamespacedName := client.ObjectKey{
+		Namespace: r.CAPINamespace,
+		Name:      req.Name,
+	}
+
+	if err := r.Get(ctx, capiNamespacedName, capiMachine); apierrors.IsNotFound(err) {
+		logger.Info("CAPI Machine not found")
+
+		capiMachineNotFound = true
+	} else if err != nil {
+		logger.Error(err, "Failed to get CAPI Machine")
+		return ctrl.Result{}, fmt.Errorf("failed to get CAPI machine:: %w", err)
+	}
+
+	if mapiMachineNotFound && capiMachineNotFound {
+		logger.Info("CAPI and MAPI machines not found, nothing to do")
+		return ctrl.Result{}, nil
+	}
+
+	// We mirror if the CAPI machine is owned by a MachineSet which has a MAPI
+	// counterpart. This is because we want to be able to migrate in both directions.
+	if mapiMachineNotFound {
+		if shouldReconcile, err := r.shouldMirrorCAPIMachineToMAPIMachine(ctx, logger, capiMachine); err != nil {
+			return ctrl.Result{}, err
+		} else if shouldReconcile {
+			return r.reconcileCAPIMachinetoMAPIMachine(ctx, capiMachine, mapiMachine)
+		}
+	}
+
+	switch mapiMachine.Status.AuthoritativeAPI {
+	case machinev1beta1.MachineAuthorityMachineAPI:
+		return r.reconcileMAPIMachinetoCAPIMachine(ctx, mapiMachine, capiMachine)
+	case machinev1beta1.MachineAuthorityClusterAPI:
+		return r.reconcileCAPIMachinetoMAPIMachine(ctx, capiMachine, mapiMachine)
+	case machinev1beta1.MachineAuthorityMigrating:
+		logger.Info("machine currently migrating", "machine", mapiMachine.GetName())
+		return ctrl.Result{}, nil
+	default:
+		return ctrl.Result{}, fmt.Errorf("%w: %s", errAuthoritativeAPIUnexpected, mapiMachine.Spec.AuthoritativeAPI)
```

Comment:
Nit, this would go into a crash loop back-off, better to log the error and return a nil to prevent the back-off loop.

That said, the API validations should mean we never get here

Reply:
TIL, would a single instance of this cause it to crash loop, even if all of the others were fine? I guess it would keep reconciling and keep erroring?

From the point of a user, if we are in a situation where we'd got past the API validations would we want this to surface as a log or as a more obvious failure? 🤔

Reply:
Sorry, wording not great there, I don't actually mean pod level crash loop back off, I meant the reconcile would go into exponential backoff 🤦 

It will log  and get more infrequent, so, this is fine, but, I guess since it's not a transient error, retrying has no way to resolve, it would be better not to bother

Reply:
Ohh ok gotcha. Makes sense now!